### DISCUSSION TITLE
방향키와 점프 컨트롤러를 껐다/켰다 하는 방법이 있을까요?

### DISCUSSION THREAD

**Discussion by TaejunPark** [naverz/zepeto-studio-kor#625] [5. Q&A]
코드로 캐릭터에게 점프를 시켰는데
방향키로 움직이고 있으면(아이들이 아니면?) 한번씩 점프가 안돼서, 방법을 찾는 중입니다.

컨트롤러를 disable/enable 하는 코드가 있을까요?  잘 못찾겟네요

시간내어 읽어주셔서 감사합니다.

**Comment by Zepeto-TechEvangelist** [comment]
안녕하세요, 혹시 사용하시는 캐릭터 컨트롤러 모션이 V1이시라면 V2로 변경하고 테스트 해보시길 권장 드립니다.

제어하는 방법에 대해 안내를 드리자면 
캐릭터 모션은 유니티의 New Input System으로 직접 제어할 수 있습니다.

Hierarchy에 있는 ZepetoPlayers > ZepetoInputControl 을 클릭하시면 
아래와 같이 New Input System으로 제작된 컨트롤에 접근 가능합니다.

<img width="516" alt="스크린샷 2022-04-18 오전 10 01 56" src="https://user-images.githubusercontent.com/100118650/163739300-8f1445d6-43c8-4d10-94d2-369e8e46f16f.png">

직접 설정을 수정하셔도 되고, 스크립트로 제어도 가능합니다.

<img width="1300" alt="스크린샷 2022-04-18 오전 10 08 15" src="https://user-images.githubusercontent.com/100118650/163739518-c8f4aff9-e435-4398-a39c-29c25c19fc32.png">

New Input  System을 이용한 이동과 점프 제어 스크립트 예제입니다.
원하시는 내용에 도움이 되시길 바라며, 캐릭터 컨트롤과 관련한 샘플 예제 및 가이드도 준비 중에 있습니다.
감사합니다. 

<hr/>


import { ZepetoScriptBehaviour } from 'ZEPETO.Script'

import {
    CharacterState, SpawnInfo, ZepetoPlayers, ZepetoPlayer,
    ZepetoCharacter, ZepetoPlayerControl, ZepetoCameraControl, ZepetoScreenTouchpad
} from 'ZEPETO.Character.Controller'

import {ZepetoInputControl} from "RootNamespace"

import {Vector2, Vector3} from "UnityEngine";

export default class UseZepetoInputControl extends ZepetoScriptBehaviour {

    public inputControl: ZepetoInputControl;

    Awake() {
        this.inputControl = new ZepetoInputControl();
    }
    Start() {
        this.inputControl.Enable();

        this.inputControl.Player.Jump.add_started((context)=>{
            console.log(`[Started] Jump - ${context.ReadValueAsObject()}`);
        });

        this.inputControl.Player.Jump.add_performed((context)=>{
            console.log(`[Performed] Jump - ${context.ReadValueAsObject()}`);
        });

        this.inputControl.Player.Jump.add_canceled((context)=>{
            console.log(`[Canceled] Jump - ${context.ReadValueAsObject()}`);
        });

        this.inputControl.Player.Move.add_started((context)=>{
            var value = context.ReadValueAsObject() as Vector2;
            console.log(`[Started] Move - ${value.ToString()}`);
        });

        this.inputControl.Player.Move.add_performed((context)=>{
            var value = context.ReadValueAsObject() as Vector2;
            console.log(`[Performed] Move - ${value.ToString()}`);
        });

        this.inputControl.Player.Move.add_canceled((context)=>{
            var value = context.ReadValueAsObject() as Vector2;
            console.log(`[Canceled] Move - ${value.ToString()}`);
        });
 }
}

**Reply by smilejsu82** [reply]
> 안녕하세요, 혹시 사용하시는 캐릭터 컨트롤러 모션이 V1이시라면 V2로 변경하고 테스트 해보시길 권장 드립니다.
> 
> 제어하는 방법에 대해 안내를 드리자면 캐릭터 모션은 유니티의 New Input System으로 직접 제어할 수 있습니다.
> 
> Hierarchy에 있는 ZepetoPlayers > ZepetoInputControl 을 클릭하시면 아래와 같이 New Input System으로 제작된 컨트롤에 접근 가능합니다.
> 
> <img alt="스크린샷 2022-04-18 오전 10 01 56" width="516" src="https://user-images.githubusercontent.com/100118650/163739300-8f1445d6-43c8-4d10-94d2-369e8e46f16f.png">
> 
> 직접 설정을 수정하셔도 되고, 스크립트로 제어도 가능합니다.
> 
> <img alt="스크린샷 2022-04-18 오전 10 08 15" width="1300" src="https://user-images.githubusercontent.com/100118650/163739518-c8f4aff9-e435-4398-a39c-29c25c19fc32.png">
> 
> New Input System을 이용한 이동과 점프 제어 스크립트 예제입니다. 원하시는 내용에 도움이 되시길 바라며, 캐릭터 컨트롤과 관련한 샘플 예제 및 가이드도 준비 중에 있습니다. 감사합니다.
> 
> import { ZepetoScriptBehaviour } from 'ZEPETO.Script'
> 
> import { CharacterState, SpawnInfo, ZepetoPlayers, ZepetoPlayer, ZepetoCharacter, ZepetoPlayerControl, ZepetoCameraControl, ZepetoScreenTouchpad } from 'ZEPETO.Character.Controller'
> 
> import {ZepetoInputControl} from "RootNamespace"
> 
> import {Vector2, Vector3} from "UnityEngine";
> 
> export default class UseZepetoInputControl extends ZepetoScriptBehaviour {
> 
> ```
> public inputControl: ZepetoInputControl;
> 
> Awake() {
>     this.inputControl = new ZepetoInputControl();
> }
> Start() {
>     this.inputControl.Enable();
> 
>     this.inputControl.Player.Jump.add_started((context)=>{
>         console.log(`[Started] Jump - ${context.ReadValueAsObject()}`);
>     });
> 
>     this.inputControl.Player.Jump.add_performed((context)=>{
>         console.log(`[Performed] Jump - ${context.ReadValueAsObject()}`);
>     });
> 
>     this.inputControl.Player.Jump.add_canceled((context)=>{
>         console.log(`[Canceled] Jump - ${context.ReadValueAsObject()}`);
>     });
> 
>     this.inputControl.Player.Move.add_started((context)=>{
>         var value = context.ReadValueAsObject() as Vector2;
>         console.log(`[Started] Move - ${value.ToString()}`);
>     });
> 
>     this.inputControl.Player.Move.add_performed((context)=>{
>         var value = context.ReadValueAsObject() as Vector2;
>         console.log(`[Performed] Move - ${value.ToString()}`);
>     });
> 
>     this.inputControl.Player.Move.add_canceled((context)=>{
>         var value = context.ReadValueAsObject() as Vector2;
>         console.log(`[Canceled] Move - ${value.ToString()}`);
>     });
> ```
> 
> } }


월드 1.2.0 에서 테스트 해봤습니다 
![image](https://user-images.githubusercontent.com/50188795/163753434-266bbe08-2e90-4d7e-a580-0d0031b5a051.png)
![image](https://user-images.githubusercontent.com/50188795/163753446-949132db-bee7-4414-82a5-bbf56eca037e.png)
![image](https://user-images.githubusercontent.com/50188795/163753465-bc4e8b76-a833-4e84-b4fd-8334ebc55e48.png)
![image](https://user-images.githubusercontent.com/50188795/163753499-d4235328-ec44-40e4-945e-afd7df35a412.png)
`import { ZepetoScriptBehaviour } from "ZEPETO.Script";

import {
  CharacterState,
  SpawnInfo,
  ZepetoPlayers,
  ZepetoPlayer,
  ZepetoCharacter,
  ZepetoPlayerControl,
  ZepetoCameraControl,
  ZepetoScreenTouchpad,
} from "ZEPETO.Character.Controller";

import { ZepetoInputControl } from "RootNamespace";

import { Vector2, Vector3 } from "UnityEngine";

export default class UseZepetoInputControl extends ZepetoScriptBehaviour {
  public inputControl: ZepetoInputControl;

  Awake() {
    this.inputControl = new ZepetoInputControl();
  }
  Start() {
    this.inputControl.Enable();

    this.inputControl.Player.Jump.add_started((context) => {
      console.log(`[Started] Jump - ${context.ReadValueAsObject()}`);
    });

    this.inputControl.Player.Jump.add_performed((context) => {
      console.log(`[Performed] Jump - ${context.ReadValueAsObject()}`);
    });

    this.inputControl.Player.Jump.add_canceled((context) => {
      console.log(`[Canceled] Jump - ${context.ReadValueAsObject()}`);
    });

    this.inputControl.Player.Move.add_started((context) => {
      var value = context.ReadValueAsObject() as Vector2;
      console.log(`[Started] Move - ${value.ToString()}`);
    });

    this.inputControl.Player.Move.add_performed((context) => {
      var value = context.ReadValueAsObject() as Vector2;
      console.log(`[Performed] Move - ${value.ToString()}`);
    });

    this.inputControl.Player.Move.add_canceled((context) => {
      var value = context.ReadValueAsObject() as Vector2;
      console.log(`[Canceled] Move - ${value.ToString()}`);
    });
  }
}
`

**Reply by Zepeto-TechEvangelist** [reply]
댓글에 써주신 코드를 동일하게 복사 붙여넣기를 했는데 저는 잘 됩니다 😅 
https://user-images.githubusercontent.com/100118650/163755208-dd4c146b-9a97-486e-8af3-d3d1e43ef7f6.mov

멀티플레이 샘플 위에서 한번 해보시겠어요?
그리고 안내 드린 내용은 자유롭게 커스텀 하는 내용에 대한 가이드입니다.

V2로 했을 때도  방향키로 움직이고 있을때 한번씩 점프가 안되는지 확인 부탁 드립니다.

**Reply by smilejsu82** [reply]
월드 1.2.0 입니다 
 https://github.com/naverz/zepeto-multiplay-example
위 멀티 플레이 샘플에서 아무것도 안건드리고 스크립트 만들어서 확인해봤는데요 
동일한 에러가 나는것을 확인 했습니다 
![image](https://user-images.githubusercontent.com/50188795/163759067-659c54da-d530-4e0b-b5a8-b1d0bdb7af50.png)
![image](https://user-images.githubusercontent.com/50188795/163759151-b2e107ed-24d3-4f5a-b63c-ccfc1e73492c.png)
. 찍어 보니깐 속성은 있는거 같은데..무엇이 문제일까요...?


context가 출력 되긴 하네요 
![image](https://user-images.githubusercontent.com/50188795/163760249-33f11625-dc4a-479b-aadb-270435dc2e02.png)
![image](https://user-images.githubusercontent.com/50188795/163760230-822a6844-0807-48df-b463-e45006a323b5.png)

근데 이부분은 왜 컴파일 에러가 나는것일까요...  ~ ,. ~ ?
![image](https://user-images.githubusercontent.com/50188795/163760556-f83bef7f-206b-4655-bf63-b566fc331428.png)
![image](https://user-images.githubusercontent.com/50188795/163760740-73a3618e-e68a-449a-9123-1b7703b342aa.png)
![image](https://user-images.githubusercontent.com/50188795/163760847-31946514-f969-4e6d-a061-5baf3ad244fa.png)

**Reply by TaejunPark** [reply]
주말중에 그냥 ui 투명하게해서 대충 터치를 막아놨었는데, 다음에 프로젝트 열때 해보겠습니다. 잘 안보던 부분인데 상세한 답변 감사합니다☺️☺️

**Comment by smilejsu82** [comment]
다음과 같이 추가 한 Action의 경우 스크립트에서 어떻게 가져와야 하나요?
![image](https://user-images.githubusercontent.com/50188795/163763526-dffab9af-e5bc-41df-9d32-0b7731060918.png)
![image](https://user-images.githubusercontent.com/50188795/163763626-18956ffc-9076-4aad-844d-d2f34d96880e.png)
